crates/boot_srv/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+///
+/// #### 3. Storage Strategies
+///
+/// ##### 3.1. The Future
+///
+/// It is the intention to someday in the future to add a "trusted" strategy,
+/// that will be triggerd via a new api, perhaps `/registerTrust/Base64Space`.
+///
+/// Even when that API is implemented, however, the default strategy defined
+/// next will be used on all spaces that have not been registered with a
+/// different strategy. Therefore, we can proceed for now without consideration
+/// for any other future strategies.
+///
+/// ##### 3.2. The "Default" Storage Strategy
+///
+/// Assumptions:
+///
+/// - We don't want to store unbounded count infos in each space.
+/// - We like storing long-running reliable nodes.
+/// - We don't want to ONLY store long-running nodes to avoid eclipse scenarios.
+///
+/// The solution put forth here is to allocate half our storage space to
+/// long-running nodes, and the other half to whoever has most recently
+/// published an info. The strategy for accomplishing this is defined here:
+///
+/// Consider "the store" acting as a stack. It can be implemented in any manner,
+/// but the following strategy assumes new entries are added to the end,
+/// and when any entries are removed, the indexes of any items after them
+/// will be decrimented.
+///
+/// - The server SHOULD provide a configurable per-space max info count.
+///   For the duration of this document, that value will be called MAX_INFOS.
+///   It is recommended to default that value to `32`.
+/// - The server SHOULD delete expired infos periodically.
+/// - If the server is already storing an agent, and a `PUT` with a newer
+///   `createdAt` arrives, the existing entry MUST be replaced.
+/// - If the store count is < MAX_INFOS, the server MUST push the info
+///   onto the stack.
+/// - If the store count is >= MAX_INFOS, the server MUST delete the entry
+///   at MAX_INFOS / 2 and then push the info onto the stack.
+///
+/// #### 4. Rate Limiting
+///
+/// Info count within a space is limited by the storage strategy above. But
+/// space count within a server is unbounded. While servers will likely want
+/// to limit request counts in general, we are also going to have to take some
+/// special care with PUT requests that result in creation of a new space that
+/// was not previously tracked in order to mitigate attacks.
+///
+/// - A server SHOULD delete any spaces that no longer contain agents.
+/// - A server SHOULD decide a max count of spaces it intends to support
+///   (based on available memory or disk space for storing agents within those
+///   spaces), and then return error 429 beyond that.
+/// - A server MAY make the above max space count configurable.
+/// - A server SHOULD track the IP addresses of clients that make PUT requests
+///   which result in space creation, and error with 429 if that frequency
+///   is beyond a limit.
+/// - A server MAY make this limit configurable.
+#[cfg(doc)]
+pub mod spec {}
```

Comment:
Can we also have a /metrics path which behaves similar to the current metrics system (showing number of space and active nodes in each space (or maybe number of trusted and self-published nodes in each space)) for a block of time, except this time have the results be REVERSE chronological showing the most recent time blocks first?

Reply:
Usually metrics are a server-specific implementation detail. I'm happy to include a metrics api with the server.

@artbrock - do you think a metrics api needs to be included in the spec? Indicating that any k2 bootstrap server implementation anybody made would respond with the exact same metrics?